src/sources/kafka.rs

```diff
@@ -236,25 +238,43 @@ async fn kafka_source(
     acknowledgements: bool,
 ) -> Result<(), ()> {
     let consumer = Arc::new(consumer);
-    let (finalizer, mut ack_stream) =
-        OrderedFinalizer::<FinalizerEntry>::maybe_new(acknowledgements, shutdown.clone());
+    let (mut finalizer, mut ack_stream) =
+        OrderedFinalizer::<FinalizerEntry>::maybe_new_without_shutdown(acknowledgements);
     let mut stream = consumer.stream();
     let keys = Keys::from(log_schema(), &config);
 
+    let mut shutting_down = false;
     loop {
         tokio::select! {
-            _ = &mut shutdown => break,
-            entry = ack_stream.next() => if let Some((status, entry)) = entry {
-                if status == BatchStatus::Delivered {
-                    if let Err(error) =
-                        consumer.store_offset(&entry.topic, entry.partition, entry.offset)
-                    {
-                        emit!(KafkaOffsetUpdateError { error });
+            _ = &mut shutdown => {
+                if let Ok(topics) = consumer.subscription() {
+                    if let Err(error) = consumer.pause(&topics) {
+                        // Might not be the best error type for this :/
+                        emit!(KafkaReadError { error });
```

Comment:
No, we probably should have another error here. @StephenWakely @neuronull do you have any thoughts here?

Reply:
Makes sense to me to have a separate error. It looks like `KafkaOffsetUpdateError` , and `KafkaReadError` are both consuming a `rdkafka::error::KafkaError` , which is the same Err returned by `consumer.pause()` , so you could model it after those.

It looks like a new event could be added, something like `KafkaPauseError` , which could have a message along the lines of `"Failed to pause consumption of the specified topics." ` (taken from the docstring of the `pause()` function), and you could even provide the `&topics` as a member of the `KafkaPauseError`.